File: leetcode/go/1356-sortIntegersByTheNumberOf1Bits.go
//go:build ignore

package main

import (
	"fmt"
	"sort"
)

func countBits(num int) int {
	cnt := 0
	for num > 0 {
		cnt += num & 1
		num >>= 1
	}
	return cnt
}

func sortByBits(arr []int) []int {
	sort.Slice(arr, func(i, j int) bool {
		aBits := countBits(arr[i])
		bBits := countBits(arr[j])

		if aBits == bBits {
			return arr[i] < arr[j]
		}
		return aBits < bBits
	})
	return arr
}

func main() {
	arr := []int{0, 1, 2, 3, 4, 5, 6, 7, 8}
	fmt.Print(sortByBits(arr))
}

File: leetcode/python/1356-sortIntegersByTheNumberOf1Bits.py
from typing import List

class Solution:
    def sortByBits(self, arr: List[int]) -> List[int]:
        tmpArr = []
        
        for num in arr:
            ones = bin(num).count('1')
            tmpArr.append((num, ones))
        
        tmpArr.sort(key=lambda x: (x[1], x[0]))
        
        return [x[0] for x in tmpArr]
    
def main():
    s = Solution()
    arr = [0, 1, 2, 3, 4, 5, 6, 7, 8]
    print(s.sortByBits(arr))

if __name__ == "__main__":
    main()
    